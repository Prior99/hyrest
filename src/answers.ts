import * as HTTP from "http-status-codes";

export interface Answer<T> {
    readonly message?: string;
    readonly data?: T;
}

export interface WrappedAnswer<T> {
    readonly statusCode: number;
    readonly result: Answer<T>;
}

export interface Wrapper<T> {
    readonly body?: T;
    readonly message?: string;

}

function isWrapper(wrapper: any): wrapper is Wrapper<any> {
    if (typeof wrapper !== "object") {
        return false;
    }
    return Object.keys(wrapper).every(key => ["message", "body"].includes(key));
}

function answer(statusCode: number, ...args: any[]): any {
    if (args.length === 0) {
        return {
            statusCode,
        };
    }
    if (args.length === 1) {
        if (isWrapper(args[0])) {
            const { message, data } = args[0];
            return {
                statusCode,
                result: {
                    message,
                    data,
                },
            };
        }
        if (typeof args[0] === "string") {
            return {
                statusCode,
                result: {
                    message: args[0],
                },
            };
        }
        return {
            statusCode,
            result: {
                data: args[0],
            },
        };
    }
    return {
        statusCode,
        result: {
            data: args[0],
            message: args[1],
        },
    };
}

export function accepted<T>(wrapper: Wrapper<T>): T;
export function accepted<T>(body?: T): T;
export function accepted<T>(message?: string): T;
export function accepted<T>(body?: T, message?: string): T;
export function accepted<T>(...args: any[]): T {
    return answer(HTTP.ACCEPTED, ...args) as T;
}

export function badGateway<T>(wrapper: Wrapper<T>): T;
export function badGateway<T>(body?: T): T;
export function badGateway<T>(message?: string): T;
export function badGateway<T>(body?: T, message?: string): T;
export function badGateway<T>(...args: any[]): T {
    return answer(HTTP.BAD_GATEWAY, ...args) as T;
}

export function badRequest<T>(wrapper: Wrapper<T>): T;
export function badRequest<T>(body?: T): T;
export function badRequest<T>(message?: string): T;
export function badRequest<T>(body?: T, message?: string): T;
export function badRequest<T>(...args: any[]): T {
    return answer(HTTP.BAD_REQUEST, ...args) as T;
}

export function conflict<T>(wrapper: Wrapper<T>): T;
export function conflict<T>(body?: T): T;
export function conflict<T>(message?: string): T;
export function conflict<T>(body?: T, message?: string): T;
export function conflict<T>(...args: any[]): T {
    return answer(HTTP.CONFLICT, ...args) as T;
}

export function httpContinue<T>(wrapper: Wrapper<T>): T;
export function httpContinue<T>(body?: T): T;
export function httpContinue<T>(message?: string): T;
export function httpContinue<T>(body?: T, message?: string): T;
export function httpContinue<T>(...args: any[]): T {
    return answer(HTTP.CONTINUE, ...args) as T;
}

export function created<T>(wrapper: Wrapper<T>): T;
export function created<T>(body?: T): T;
export function created<T>(message?: string): T;
export function created<T>(body?: T, message?: string): T;
export function created<T>(...args: any[]): T {
    return answer(HTTP.CREATED, ...args) as T;
}

export function expectationFailed<T>(wrapper: Wrapper<T>): T;
export function expectationFailed<T>(body?: T): T;
export function expectationFailed<T>(message?: string): T;
export function expectationFailed<T>(body?: T, message?: string): T;
export function expectationFailed<T>(...args: any[]): T {
    return answer(HTTP.EXPECTATION_FAILED, ...args) as T;
}

export function failedDependency<T>(wrapper: Wrapper<T>): T;
export function failedDependency<T>(body?: T): T;
export function failedDependency<T>(message?: string): T;
export function failedDependency<T>(body?: T, message?: string): T;
export function failedDependency<T>(...args: any[]): T {
    return answer(HTTP.FAILED_DEPENDENCY, ...args) as T;
}

export function forbidden<T>(wrapper: Wrapper<T>): T;
export function forbidden<T>(body?: T): T;
export function forbidden<T>(message?: string): T;
export function forbidden<T>(body?: T, message?: string): T;
export function forbidden<T>(...args: any[]): T {
    return answer(HTTP.FORBIDDEN, ...args) as T;
}

export function gatewayTimeout<T>(wrapper: Wrapper<T>): T;
export function gatewayTimeout<T>(body?: T): T;
export function gatewayTimeout<T>(message?: string): T;
export function gatewayTimeout<T>(body?: T, message?: string): T;
export function gatewayTimeout<T>(...args: any[]): T {
    return answer(HTTP.GATEWAY_TIMEOUT, ...args) as T;
}

export function gone<T>(wrapper: Wrapper<T>): T;
export function gone<T>(body?: T): T;
export function gone<T>(message?: string): T;
export function gone<T>(body?: T, message?: string): T;
export function gone<T>(...args: any[]): T {
    return answer(HTTP.GONE, ...args) as T;
}

export function httpVersionNotSupported<T>(wrapper: Wrapper<T>): T;
export function httpVersionNotSupported<T>(body?: T): T;
export function httpVersionNotSupported<T>(message?: string): T;
export function httpVersionNotSupported<T>(body?: T, message?: string): T;
export function httpVersionNotSupported<T>(...args: any[]): T {
    return answer(HTTP.HTTP_VERSION_NOT_SUPPORTED, ...args) as T;
}

export function insufficientSpaceOnResource<T>(wrapper: Wrapper<T>): T;
export function insufficientSpaceOnResource<T>(body?: T): T;
export function insufficientSpaceOnResource<T>(message?: string): T;
export function insufficientSpaceOnResource<T>(body?: T, message?: string): T;
export function insufficientSpaceOnResource<T>(...args: any[]): T {
    return answer(HTTP.INSUFFICIENT_SPACE_ON_RESOURCE, ...args) as T;
}

export function insufficientStorage<T>(wrapper: Wrapper<T>): T;
export function insufficientStorage<T>(body?: T): T;
export function insufficientStorage<T>(message?: string): T;
export function insufficientStorage<T>(body?: T, message?: string): T;
export function insufficientStorage<T>(...args: any[]): T {
    return answer(HTTP.INSUFFICIENT_STORAGE, ...args) as T;
}

export function internalServerError<T>(wrapper: Wrapper<T>): T;
export function internalServerError<T>(body?: T): T;
export function internalServerError<T>(message?: string): T;
export function internalServerError<T>(body?: T, message?: string): T;
export function internalServerError<T>(...args: any[]): T {
    return answer(HTTP.INTERNAL_SERVER_ERROR, ...args) as T;
}

export function lengthRequired<T>(wrapper: Wrapper<T>): T;
export function lengthRequired<T>(body?: T): T;
export function lengthRequired<T>(message?: string): T;
export function lengthRequired<T>(body?: T, message?: string): T;
export function lengthRequired<T>(...args: any[]): T {
    return answer(HTTP.LENGTH_REQUIRED, ...args) as T;
}

export function locked<T>(wrapper: Wrapper<T>): T;
export function locked<T>(body?: T): T;
export function locked<T>(message?: string): T;
export function locked<T>(body?: T, message?: string): T;
export function locked<T>(...args: any[]): T {
    return answer(HTTP.LOCKED, ...args) as T;
}

export function methodFailure<T>(wrapper: Wrapper<T>): T;
export function methodFailure<T>(body?: T): T;
export function methodFailure<T>(message?: string): T;
export function methodFailure<T>(body?: T, message?: string): T;
export function methodFailure<T>(...args: any[]): T {
    return answer(HTTP.METHOD_FAILURE, ...args) as T;
}

export function methodNotAllowed<T>(wrapper: Wrapper<T>): T;
export function methodNotAllowed<T>(body?: T): T;
export function methodNotAllowed<T>(message?: string): T;
export function methodNotAllowed<T>(body?: T, message?: string): T;
export function methodNotAllowed<T>(...args: any[]): T {
    return answer(HTTP.METHOD_NOT_ALLOWED, ...args) as T;
}

export function movedPermanently<T>(wrapper: Wrapper<T>): T;
export function movedPermanently<T>(body?: T): T;
export function movedPermanently<T>(message?: string): T;
export function movedPermanently<T>(body?: T, message?: string): T;
export function movedPermanently<T>(...args: any[]): T {
    return answer(HTTP.MOVED_PERMANENTLY, ...args) as T;
}

export function movedTemporarily<T>(wrapper: Wrapper<T>): T;
export function movedTemporarily<T>(body?: T): T;
export function movedTemporarily<T>(message?: string): T;
export function movedTemporarily<T>(body?: T, message?: string): T;
export function movedTemporarily<T>(...args: any[]): T {
    return answer(HTTP.MOVED_TEMPORARILY, ...args) as T;
}

export function multiStatus<T>(wrapper: Wrapper<T>): T;
export function multiStatus<T>(body?: T): T;
export function multiStatus<T>(message?: string): T;
export function multiStatus<T>(body?: T, message?: string): T;
export function multiStatus<T>(...args: any[]): T {
    return answer(HTTP.MULTI_STATUS, ...args) as T;
}

export function multipleChoices<T>(wrapper: Wrapper<T>): T;
export function multipleChoices<T>(body?: T): T;
export function multipleChoices<T>(message?: string): T;
export function multipleChoices<T>(body?: T, message?: string): T;
export function multipleChoices<T>(...args: any[]): T {
    return answer(HTTP.MULTIPLE_CHOICES, ...args) as T;
}

export function networkAuthenticationRequired<T>(wrapper: Wrapper<T>): T;
export function networkAuthenticationRequired<T>(body?: T): T;
export function networkAuthenticationRequired<T>(message?: string): T;
export function networkAuthenticationRequired<T>(body?: T, message?: string): T;
export function networkAuthenticationRequired<T>(...args: any[]): T {
    return answer(HTTP.NETWORK_AUTHENTICATION_REQUIRED, ...args) as T;
}

export function noContent<T>(wrapper: Wrapper<T>): T;
export function noContent<T>(body?: T): T;
export function noContent<T>(message?: string): T;
export function noContent<T>(body?: T, message?: string): T;
export function noContent<T>(...args: any[]): T {
    return answer(HTTP.NO_CONTENT, ...args) as T;
}

export function nonAuthoritativeInformation<T>(wrapper: Wrapper<T>): T;
export function nonAuthoritativeInformation<T>(body?: T): T;
export function nonAuthoritativeInformation<T>(message?: string): T;
export function nonAuthoritativeInformation<T>(body?: T, message?: string): T;
export function nonAuthoritativeInformation<T>(...args: any[]): T {
    return answer(HTTP.NON_AUTHORITATIVE_INFORMATION, ...args) as T;
}

export function notAcceptable<T>(wrapper: Wrapper<T>): T;
export function notAcceptable<T>(body?: T): T;
export function notAcceptable<T>(message?: string): T;
export function notAcceptable<T>(body?: T, message?: string): T;
export function notAcceptable<T>(...args: any[]): T {
    return answer(HTTP.NOT_ACCEPTABLE, ...args) as T;
}

export function notFound<T>(wrapper: Wrapper<T>): T;
export function notFound<T>(body?: T): T;
export function notFound<T>(message?: string): T;
export function notFound<T>(body?: T, message?: string): T;
export function notFound<T>(...args: any[]): T {
    return answer(HTTP.NOT_FOUND, ...args) as T;
}

export function notImplemented<T>(wrapper: Wrapper<T>): T;
export function notImplemented<T>(body?: T): T;
export function notImplemented<T>(message?: string): T;
export function notImplemented<T>(body?: T, message?: string): T;
export function notImplemented<T>(...args: any[]): T {
    return answer(HTTP.NOT_IMPLEMENTED, ...args) as T;
}

export function notModified<T>(wrapper: Wrapper<T>): T;
export function notModified<T>(body?: T): T;
export function notModified<T>(message?: string): T;
export function notModified<T>(body?: T, message?: string): T;
export function notModified<T>(...args: any[]): T {
    return answer(HTTP.NOT_MODIFIED, ...args) as T;
}

export function ok<T>(wrapper: Wrapper<T>): T;
export function ok<T>(body?: T): T;
export function ok<T>(message?: string): T;
export function ok<T>(body?: T, message?: string): T;
export function ok<T>(...args: any[]): T {
    return answer(HTTP.OK, ...args) as T;
}

export function partialContent<T>(wrapper: Wrapper<T>): T;
export function partialContent<T>(body?: T): T;
export function partialContent<T>(message?: string): T;
export function partialContent<T>(body?: T, message?: string): T;
export function partialContent<T>(...args: any[]): T {
    return answer(HTTP.PARTIAL_CONTENT, ...args) as T;
}

export function paymentRequired<T>(wrapper: Wrapper<T>): T;
export function paymentRequired<T>(body?: T): T;
export function paymentRequired<T>(message?: string): T;
export function paymentRequired<T>(body?: T, message?: string): T;
export function paymentRequired<T>(...args: any[]): T {
    return answer(HTTP.PAYMENT_REQUIRED, ...args) as T;
}

export function preconditionFailed<T>(wrapper: Wrapper<T>): T;
export function preconditionFailed<T>(body?: T): T;
export function preconditionFailed<T>(message?: string): T;
export function preconditionFailed<T>(body?: T, message?: string): T;
export function preconditionFailed<T>(...args: any[]): T {
    return answer(HTTP.PRECONDITION_FAILED, ...args) as T;
}

export function preconditionRequired<T>(wrapper: Wrapper<T>): T;
export function preconditionRequired<T>(body?: T): T;
export function preconditionRequired<T>(message?: string): T;
export function preconditionRequired<T>(body?: T, message?: string): T;
export function preconditionRequired<T>(...args: any[]): T {
    return answer(HTTP.PRECONDITION_REQUIRED, ...args) as T;
}

export function processing<T>(wrapper: Wrapper<T>): T;
export function processing<T>(body?: T): T;
export function processing<T>(message?: string): T;
export function processing<T>(body?: T, message?: string): T;
export function processing<T>(...args: any[]): T {
    return answer(HTTP.PROCESSING, ...args) as T;
}

export function proxyAuthenticationRequired<T>(wrapper: Wrapper<T>): T;
export function proxyAuthenticationRequired<T>(body?: T): T;
export function proxyAuthenticationRequired<T>(message?: string): T;
export function proxyAuthenticationRequired<T>(body?: T, message?: string): T;
export function proxyAuthenticationRequired<T>(...args: any[]): T {
    return answer(HTTP.PROXY_AUTHENTICATION_REQUIRED, ...args) as T;
}

export function requestHeaderFieldsTooLarge<T>(wrapper: Wrapper<T>): T;
export function requestHeaderFieldsTooLarge<T>(body?: T): T;
export function requestHeaderFieldsTooLarge<T>(message?: string): T;
export function requestHeaderFieldsTooLarge<T>(body?: T, message?: string): T;
export function requestHeaderFieldsTooLarge<T>(...args: any[]): T {
    return answer(HTTP.REQUEST_HEADER_FIELDS_TOO_LARGE, ...args) as T;
}

export function requestTimeout<T>(wrapper: Wrapper<T>): T;
export function requestTimeout<T>(body?: T): T;
export function requestTimeout<T>(message?: string): T;
export function requestTimeout<T>(body?: T, message?: string): T;
export function requestTimeout<T>(...args: any[]): T {
    return answer(HTTP.REQUEST_TIMEOUT, ...args) as T;
}

export function requestTooLong<T>(wrapper: Wrapper<T>): T;
export function requestTooLong<T>(body?: T): T;
export function requestTooLong<T>(message?: string): T;
export function requestTooLong<T>(body?: T, message?: string): T;
export function requestTooLong<T>(...args: any[]): T {
    return answer(HTTP.REQUEST_TOO_LONG, ...args) as T;
}

export function requestUriTooLong<T>(wrapper: Wrapper<T>): T;
export function requestUriTooLong<T>(body?: T): T;
export function requestUriTooLong<T>(message?: string): T;
export function requestUriTooLong<T>(body?: T, message?: string): T;
export function requestUriTooLong<T>(...args: any[]): T {
    return answer(HTTP.REQUEST_URI_TOO_LONG, ...args) as T;
}

export function requestedRangeNotSatisfiable<T>(wrapper: Wrapper<T>): T;
export function requestedRangeNotSatisfiable<T>(body?: T): T;
export function requestedRangeNotSatisfiable<T>(message?: string): T;
export function requestedRangeNotSatisfiable<T>(body?: T, message?: string): T;
export function requestedRangeNotSatisfiable<T>(...args: any[]): T {
    return answer(HTTP.REQUESTED_RANGE_NOT_SATISFIABLE, ...args) as T;
}

export function resetContent<T>(wrapper: Wrapper<T>): T;
export function resetContent<T>(body?: T): T;
export function resetContent<T>(message?: string): T;
export function resetContent<T>(body?: T, message?: string): T;
export function resetContent<T>(...args: any[]): T {
    return answer(HTTP.RESET_CONTENT, ...args) as T;
}

export function seeOther<T>(wrapper: Wrapper<T>): T;
export function seeOther<T>(body?: T): T;
export function seeOther<T>(message?: string): T;
export function seeOther<T>(body?: T, message?: string): T;
export function seeOther<T>(...args: any[]): T {
    return answer(HTTP.SEE_OTHER, ...args) as T;
}

export function serviceUnavailable<T>(wrapper: Wrapper<T>): T;
export function serviceUnavailable<T>(body?: T): T;
export function serviceUnavailable<T>(message?: string): T;
export function serviceUnavailable<T>(body?: T, message?: string): T;
export function serviceUnavailable<T>(...args: any[]): T {
    return answer(HTTP.SERVICE_UNAVAILABLE, ...args) as T;
}

export function switchingProtocols<T>(wrapper: Wrapper<T>): T;
export function switchingProtocols<T>(body?: T): T;
export function switchingProtocols<T>(message?: string): T;
export function switchingProtocols<T>(body?: T, message?: string): T;
export function switchingProtocols<T>(...args: any[]): T {
    return answer(HTTP.SWITCHING_PROTOCOLS, ...args) as T;
}

export function temporaryRedirect<T>(wrapper: Wrapper<T>): T;
export function temporaryRedirect<T>(body?: T): T;
export function temporaryRedirect<T>(message?: string): T;
export function temporaryRedirect<T>(body?: T, message?: string): T;
export function temporaryRedirect<T>(...args: any[]): T {
    return answer(HTTP.TEMPORARY_REDIRECT, ...args) as T;
}

export function tooManyRequests<T>(wrapper: Wrapper<T>): T;
export function tooManyRequests<T>(body?: T): T;
export function tooManyRequests<T>(message?: string): T;
export function tooManyRequests<T>(body?: T, message?: string): T;
export function tooManyRequests<T>(...args: any[]): T {
    return answer(HTTP.TOO_MANY_REQUESTS, ...args) as T;
}

export function unauthorized<T>(wrapper: Wrapper<T>): T;
export function unauthorized<T>(body?: T): T;
export function unauthorized<T>(message?: string): T;
export function unauthorized<T>(body?: T, message?: string): T;
export function unauthorized<T>(...args: any[]): T {
    return answer(HTTP.UNAUTHORIZED, ...args) as T;
}

export function unprocessableEntity<T>(wrapper: Wrapper<T>): T;
export function unprocessableEntity<T>(body?: T): T;
export function unprocessableEntity<T>(message?: string): T;
export function unprocessableEntity<T>(body?: T, message?: string): T;
export function unprocessableEntity<T>(...args: any[]): T {
    return answer(HTTP.UNPROCESSABLE_ENTITY, ...args) as T;
}

export function unsupportedMediaType<T>(wrapper: Wrapper<T>): T;
export function unsupportedMediaType<T>(body?: T): T;
export function unsupportedMediaType<T>(message?: string): T;
export function unsupportedMediaType<T>(body?: T, message?: string): T;
export function unsupportedMediaType<T>(...args: any[]): T {
    return answer(HTTP.UNSUPPORTED_MEDIA_TYPE, ...args) as T;
}

export function useProxy<T>(wrapper: Wrapper<T>): T;
export function useProxy<T>(body?: T): T;
export function useProxy<T>(message?: string): T;
export function useProxy<T>(body?: T, message?: string): T;
export function useProxy<T>(...args: any[]): T {
    return answer(HTTP.USE_PROXY, ...args) as T;
}
