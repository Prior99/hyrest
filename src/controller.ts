import { Constructable } from "./types";
import "isomorphic-fetch";
import "reflect-metadata";

import { Route } from "./route";
import { Params, ApiError, HTTPMethod } from "./types";
import { compile } from "path-to-regexp";
import { isBrowser } from "./is-browser";
import { populate } from "./scope";

export enum ControllerMode {
    SERVER = "server",
    CLIENT = "client",
}

export function getDefaultControllerMode() {
    return isBrowser() ? ControllerMode.CLIENT : ControllerMode.SERVER;
}

function buildQueryString(query: Params) {
    const queryString = Object.keys(query).reduce((result, key) => {
        const unencoded = query[key];
        if (typeof unencoded === "undefined" || unencoded === null) {
            return result;
        }
        const value = encodeURIComponent(unencoded);
        return `${result}&${key}=${value}`;
    }, "");
    // Remove first `&`.
    return queryString.substr(1, queryString.length);
}

/**
 * Assembles a full Url from an express url template, url parameters, as base Url and a query parameters.
 *
 * @param urlParameters The parameters fed to the express url template specified in `subUrl`.
 * @param query The query parameters for the url.
 * @param baseUrl The url to which the template url in `subUrl` is relative to.
 * @param subUrl An url template as used in express.
 *
 * @return The assembled Url.
 */
export function buildUrl(urlParameters: Params, query: Params, baseUrl: string, subUrl: string) {
    const queryString = buildQueryString(query);
    // The url parameters need to be explicitly converted to string as they will be transported via
    // the url string.
    const stringifiedParameters = Object.keys(urlParameters).reduce((result, key) => {
        result[key] = `${urlParameters[key]}`;
        return result;
    }, {} as any);

    // Compile the route's pattern (This is the reverse of what express does with url patterns) and
    // feed it the url parameters which had been converted to string.
    const routeString = compile(subUrl)(stringifiedParameters);

    // Assemble the full url.
    return `${baseUrl}${routeString}?${queryString}`;
}

/**
 * A handler which will be called whenever an Api error is encountered.
 *
 * @param error The Api error that was encountered.
 */
export type ErrorHandler = (error: ApiError) => void;

/**
 * Options with which a controller can be configured.
 * The options are used for all instances of a Controller class.
 */
export interface ControllerOptions {
    /**
     * If set to `true` the controller will `throw` whenever an error is encountered.
     * Default is `true`.
     */
    readonly throwOnError?: boolean;
    /**
     * When supplied, this handler will be called for all errors encountered by this controller.
     */
    readonly errorHandler?: ErrorHandler;
    /**
     * A base Url on which the backend can be publicly reached.
     */
    readonly baseUrl?: string;
    /**
     * Can be `ControllerMode.CLIENT` or `ControllerMode.SERVER`. This is autodetected by default
     * and `ControllerMode.CLIENT` will be used in the browser. This determines if the controller
     * will be used to serve the Api or consume it.
     */
    readonly mode?: ControllerMode;
}

export class Controller {
    public throwOnError = true;
    public errorHandler: ErrorHandler;
    public baseUrl: string;
    public mode: ControllerMode = getDefaultControllerMode();

    constructor(options: ControllerOptions) {
        if (options) {
            this.configure(options);
        }
    }

    /**
     * Will be called by `configureController` and applies the given options to this controller.
     */
    public configure(options: ControllerOptions) {
        const { throwOnError, errorHandler, baseUrl, mode } = options;
        if (typeof throwOnError !== "undefined") { this.throwOnError = throwOnError; }
        if (typeof errorHandler !== "undefined") { this.errorHandler = errorHandler; }
        if (typeof baseUrl !== "undefined") { this.baseUrl = baseUrl; }
        if (typeof mode !== "undefined") { this.mode = mode; }
    }

    /**
     * Will be called instead of each route's body when running in `CLIENT` mode. This function
     * will perform the `fetch` against the backend endpoint.
     *
     * @param route The Route meta information for the method for which the functions was called.
     * @param parameters The Url parameters.
     * @param body The body.
     * @param query The query parameters.
     *
     * @return The unwrapped answer from the backend.
     */
    public async wrappedFetch<T>(route: Route, urlParameters: Params, body: any, query: Params): Promise<T> {
        try {
            // Create the url from the given inputs.
            const url = buildUrl(urlParameters, query, this.baseUrl, route.url);

            // Prepare the headers.
            const headers = new Headers();
            headers.append("content-type", "application/json");

            // Perform the actual fetch.
            const response = await fetch(url, {
                body: JSON.stringify(body),
                headers,
                method: route.method,
            });

            // Parse the response from the Api as JSON.
            const answer = await response.json();

            // If the response's status code was a 2xx status code, return the response as it succeeded.
            if (response.ok) {
                const routeData = answer.data;
                if (typeof route.returnType !== "undefined" && typeof route.scope !== "undefined") {
                    if (Array.isArray(routeData)) {
                        return populate(route.scope, Array, route.returnType, routeData);
                    }
                    return populate(route.scope, route.returnType, routeData);
                }
                return routeData;
            }

            // Otherwise, create an error and handle it according to the controller's configuration.
            // This happens by throwing it, which will raise it to the `catch` block below.
            // That block will then perform the actual error handling.
            const error = new ApiError(response.status, answer);
            throw error;
        } catch (error) {
            // Error's might occur as the network might be interrupted or the response might be malformed.
            // Also an error might be thrown by the `try` block above, if a non-2xx status code has been received.
            if (this.errorHandler) {
                this.errorHandler(error);
            }
            if (this.throwOnError) {
                throw error;
            }
        }
    }
}

export function controller(options?: ControllerOptions): ClassDecorator;
export function controller<U, T extends Constructable<U>>(target: T): T;
/**
 * Decorate an individual class as an Api controller.
 *
 * @param options Global options with which this controller can be configured. `configureController` can also set
 *                set and override these options.
 *
 * @return The decorated class.
 */
export function controller<T extends Function>(arg1: ControllerOptions | T): ClassDecorator | T {
    let decorator: (target: T) => T;
    const options = typeof arg1 === "object" ? arg1 : undefined;
    decorator = function(target: T): T {
        Reflect.defineMetadata("api:controller", new Controller(options), target);
        return target;
    };
    if (typeof arg1 === "function") {
        return decorator(arg1 as T);
    }
    return decorator as ClassDecorator;
}
