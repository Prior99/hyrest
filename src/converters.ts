import { FullValidator, Processed } from "./validation";

export interface Converted<T> {
    error?: string;
    value?: T;
}

export type Converter<T> = (input: any) => Converted<T> | Promise<Converted<T>> | Processed<T> | Promise<Processed<T>>;

/**
 * Converts the given input to an integer if possible.
 *
 * @param input The input to convert to an integer. Can be anything in any form.
 *
 * @return An integer if the conversion succeeded or an error if the input was not a valid
 *         integer.
 */
export function int(input: any): Converted<number> {
    if (typeof input === "undefined") { return { value: input }; }
    const value = parseInt(input);
    if (isNaN(value) || parseFloat(input) !== value) { return { error: "Not a valid integer." }; }
    return { value };
}

/**
 * Converts the given input to a float if possible.
 *
 * @param input The input to convert to a float. Can be anything in any form.
 *
 * @return A float if the conversion succeeded or an error if the input was not a valid
 *         float.
 */
export function float(input: any): Converted<number> {
    if (typeof input === "undefined") { return { value: input }; }
    const value = parseFloat(input);
    if (isNaN(value)) { return { error: "Not a valid float." }; }
    return { value };
}

/**
 * Makes sure the given input is a string.
 *
 * @param input The input to check.
 *
 * @return The string in `input` if it was a string and an error otherwise.
 */
export function str(value: any): Converted<string> {
    if (typeof value === "undefined") { return { value }; }
    if (typeof value !== "string") { return { error: "Not a valid string." }; }
    return { value };
}

/**
 * Makes sure the given input is an object.
 *
 * @param input The input to check.
 *
 * @return The input if it was an object and an error otherwise.
 */
export function obj(value: any): Converted<Object> {
    if (typeof value === "undefined") { return { value }; }
    if (typeof value !== "object" || Array.isArray(value)) { return { error: "Not a valid object."}; }
    return { value };
}

/**
 * Makes sure the given input is a boolean.
 *
 * @param input The input to check.
 *
 * @return The boolean in `input` if it was a boolean and an error otherwise.
 */
export function bool(value: any): Converted<boolean> {
    if (typeof value === "undefined") { return { value }; }
    if (value === "true" || value === "false") { return { value: value === "true" }; }
    if (typeof value !== "boolean") { return { error: "Not a valid boolean." }; }
    return { value };
}
