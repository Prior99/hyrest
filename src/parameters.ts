export interface Params {
    [key: string]: string;
}
