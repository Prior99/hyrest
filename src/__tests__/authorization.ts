import { hyrest } from "../middleware";
import { controller, ControllerMode } from "../controller";
import { route } from "../route";
import { body, param, query } from "../parameters";
import { float, int, str } from "../converters";
import { required, email, length, only } from "../validators";
import { is } from "../validation";
import { transform } from "../transform";
import { ok, created } from "../answers";
import { createScope, scope, specify } from "../scope";
import * as request from "supertest";
import * as Express from "express";
import * as BodyParser from "body-parser";
import { authorized, unauthorized, AuthorizationMode } from "../";

test("A hyrest middleware with authorization enabled on the route", async () => {
    const mock = jest.fn();
    const ctx = {};

    @controller({ mode: ControllerMode.SERVER })
    class TestController {
        @route("GET", "/test") @authorized
        public method() { mock(); return ok(); }
    }

    const http = Express();
    http.use(BodyParser.json());
    http.use(
        hyrest(new TestController())
            .authorization((req, context) => {
                expect(context).toBe(ctx);
                return req.query["ok"] === "true";
            })
            .context(ctx),
    );

    await request(http).get("/test?ok=false").expect(401);
    expect(mock).not.toHaveBeenCalled();
    await request(http).get("/test?ok=true").expect(200);
    expect(mock).toHaveBeenCalled();
});

test("A hyrest middleware with authorization enabled on the controller", async () => {
    const mock = jest.fn();

    @authorized
    @controller({ mode: ControllerMode.SERVER })
    class TestController {
        @route("GET", "/test")
        public method() { mock(); return ok(); }
    }

    const http = Express();
    http.use(BodyParser.json());
    http.use(hyrest(new TestController()).authorization(req => req.query["ok"] === "true"));

    await request(http).get("/test?ok=false").expect(401);
    expect(mock).not.toHaveBeenCalled();
    await request(http).get("/test?ok=true").expect(200);
    expect(mock).toHaveBeenCalled();
});

test("A hyrest middleware with authorization enabled on the middleware", async () => {
    const mock = jest.fn();

    @controller({ mode: ControllerMode.SERVER })
    class TestController {
        @route("GET", "/test")
        public method() { mock(); return ok(); }
    }

    const http = Express();
    http.use(BodyParser.json());
    http.use(
        hyrest(new TestController())
            .authorization(req => req.query["ok"] === "true")
            .defaultAuthorizationMode(AuthorizationMode.AUTHORIZED),
    );

    await request(http).get("/test?ok=false").expect(401);
    expect(mock).not.toHaveBeenCalled();
    await request(http).get("/test?ok=true").expect(200);
    expect(mock).toHaveBeenCalled();
});

test("A hyrest middleware with a special authorization check", async () => {
    const mock = jest.fn();

    @authorized
    @controller({ mode: ControllerMode.SERVER })
    class TestController {
        @route("GET", "/test") @authorized({ check: req => req.query["extra"] === "true" })
        public method() { mock(); return ok(); }
    }

    const http = Express();
    http.use(BodyParser.json());
    http.use(hyrest(new TestController()).authorization(req => req.query["ok"] === "true"));

    await request(http).get("/test?ok=true").expect(401);
    expect(mock).not.toHaveBeenCalled();
    await request(http).get("/test?ok=true&extra=true").expect(200);
    expect(mock).toHaveBeenCalled();
});

test("A hyrest middleware with authorization disabled on a specific route", async () => {
    const mock = jest.fn();

    @controller({ mode: ControllerMode.SERVER })
    class TestController {
        @route("GET", "/test") @unauthorized()
        public method() { mock(); return ok(); }
    }

    const http = Express();
    http.use(BodyParser.json());
    http.use(
        hyrest(new TestController())
            .authorization(req => req.query["ok"] === "true")
            .defaultAuthorizationMode(AuthorizationMode.AUTHORIZED),
    );

    await request(http).get("/test?ok=false").expect(200);
    expect(mock).toHaveBeenCalled();
});

test("A hyrest middleware throws an error when invoked with an authorized route but no checker provided", async () => {
    const mock = jest.fn();
    const mockError = jest.fn();

    @controller({ mode: ControllerMode.SERVER })
    class TestController {
        @route("GET", "/test")
        public method() { mock(); return ok(); }
    }

    const http = Express();
    http.use(BodyParser.json());
    http.use(async (_, res: any, next: any) => {
        try {
            await next();
        } catch (err) {
            mockError(err);
            res.status(500).send();
        }
    });
    http.use(
        hyrest(new TestController()).defaultAuthorizationMode(AuthorizationMode.AUTHORIZED),
    );
    const req = await request(http)
        .get("/test?ok=false")
        .expect(500)
        .send();
    expect(mock).not.toHaveBeenCalled();
    expect(mockError.mock.calls).toMatchSnapshot();
});
