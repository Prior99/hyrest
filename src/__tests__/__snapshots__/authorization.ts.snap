// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A hyrest middleware throws an error when invoked with an authorized route but no checker provided 1`] = `
Array [
  Array [
    [Error: Call to an authorized route but no authorization check was provided.],
  ],
]
`;
